File: Algorithms/Searching/linear-search.py
import random

# Linear Search Algorithm
def linear_search(collection, target):

    # Iterate thorough all items in the collection
    for i in range(len(collection)):

        # Check if the item has the same value as target
        if collection[i] == target:
            return i 

    # Target not found
    return None


if __name__ == "__main__":
    GREEN = '\033[92m'
    RESET = '\033[0m'
    print(f"{GREEN}Linear Search{RESET}")
    
    # Create random collection of items
    data = [random.randint(1, 40) for i in range(10)]

    print (data)

    # Get the target
    target = int(input("Enter data to find : "))

    # Call linear search
    result = linear_search(data, target)

    # Check if data was found
    if result is None:
        print("Data not found")
    else:
        print(f"Found at index : {result}")

File: Algorithms/Searching/linear-search.c
#include<stdio.h>

/*
   ========================
           Linear
   ========================
*/

// Linear Search Iterative
int linear(double arr[], int n, double value) {

    for (int i = 0; i < n; i++)
        if (arr[i] == value)
            return i; // Element found

    return -1;
}

// Linear Search Recursive
int linearRec(double arr[], int n, double value) {
    if (n == 0)
        return -1; // Element not found
    else if (arr[n-1] == value)
        return n-1; // Element found
    else
        return linearRec(arr, n-1, value); // Search next index
}

int main() {

    double arr[] = {10, 20, 30, 40, 50};

    // Linear Search
    int index = linear(arr, sizeof(arr)/sizeof(arr[0]), 30);
    printf("Linear Search : %d\n", index);
    printf("Linear Search Recursive : %d\n", linearRec(arr, sizeof(arr)/sizeof(arr[0]), 30));

    return 0;
}
